### ISSUE TITLE
POS Customer field is hanging [Critical]

### ISSUE THREAD

**Issue by Olawale1** [frappe/erpnext#10591]
Hi,

The Customer field in the POS has become very slow in responding when entering a Customer Name (which is the first thing that must be done before selecting Items). As you type the name, it takes a while for the letters to even show in the field! As you can imagine, this is not acceptable in a retail environment. We currently have approx 8,000 customers on the system

Please advise 

Thanks

**Comment by tundebabzy** [comment]
Hi @Olawale1 
Can you please check your db's slow query logs?

**Comment by tundebabzy** [comment]
Hi @Olawale1 
Can you please check your db's slow query logs?

**Comment by rohitwaghchaure** [comment]
Hi @Olawale1 

We are currently working on making offline optional for POS which will release soon, in new design we have added the link field of the customer which will resolve your issue

Adding an item in **cart**
![online_pos](https://user-images.githubusercontent.com/8780500/29840434-a04a2cf8-8d1f-11e7-98f8-793001631575.gif)
 
**For payment**
![payment_for_pos](https://user-images.githubusercontent.com/8780500/29840486-d52eee04-8d1f-11e7-9739-d9bbbbeb6ebd.gif)

**Comment by Olawale1** [comment]
Hi @rohitwaghchaure 

Thanks for you response but what if we want the POS to work offline?? It appears the solution you're proposing is only for those who choose to use the POS online only... we do not want this

Please advise asap

Thanks

**Comment by rohitwaghchaure** [comment]
Hi @Olawale1 

We have did some changes in the code which we'll release soon. As we don't have large customer in the local account we're not able to replicate the issue on local account.

**Comment by Olawale1** [comment]
Hi @rohitwaghchaure 

I appreciate this a great deal. I'll update once it's released and let you know how it goes

Thanks again

**Comment by rohitwaghchaure** [comment]
Hi @Olawale1 

We have released the fix, can you please test it on your account and share the feedback. Feel free to reopen the issue, if issue still persists.

**Comment by Olawale1** [comment]
Hi @rohitwaghchaure 

I've updated to the latest version (ERPNext: v8.10.1) but not much improvement. We have a lot more serial numbers in our system than customers but there's no issue with serial numbers. Any ideas what else can be done please?

Kind regards,

**Comment by Olawale1** [comment]
Hi,

We really need to do something about this urgently. It's so critical that we've started considering deleting some of our customer records (which isn't something we'd like to do at all!)

Any suggestions please?

Thanks

**Comment by tundebabzy** [comment]
Hi. So sorry about this. Hanging sounds like a slow query. Can you check your slow logs to see of this is the case

**Comment by Olawale1** [comment]
Hi @tundebabzy 

Thanks for your response. There's however nothing in the slow query logs. I doubt it's a generic slow query issue because it works fine when online (i.e. in a regular Sales Invoice form). The issue appears to be from the POS in offline mode (working locally in the browser). That's why I was asking how the POS handles serials because we have at least 3 times the number of customers in our serial number list yet we are not experiencing any lag in offline management of serial numbers

Thanks

**Comment by Olawale1** [comment]
Any feedback please? Also, can this be re-opened since it's not actually fixed yet or do I need to create another issue?

Thanks

**Comment by rohitwaghchaure** [comment]
Hi @Olawale1 

Sorry for the issue, can it be possible for you to share the credentials of your account? Kindly PM me the application URL and the login credentials on the discuss forum. It will be helpful for us to debug the issue on your database.

**Comment by Olawale1** [comment]
Hi @rohitwaghchaure 

Thanks for your response. I've sent the requested details. It just occurred to us that deleting customer records is not even an option because they are linked to Sales Invoices so we really need and appreciate the assistance right now

Thanks

**Comment by Olawale1** [comment]
Hi @rohitwaghchaure 

Trust you're doing great. The latest update has significantly improved responsiveness but now we are facing an issue where the field only returns typed letters instead of selected value. Let's say for example I am searching for customer Rohit. When I type 'ro', I see options Robert, Rohit, Roland etc in the dropdown but when I click on Rohit, I only see 'ro' in the customer field! I believe it's the same issue that was reported in the issue below:

https://github.com/frappe/erpnext/issues/10705

Kindly help fix

Thanks

**Comment by rohitwaghchaure** [comment]
Hi @Olawale1 

We have fixed the issue, can you please check. I am closing the issue, feel free to reopen the issue if it's still persists

**Comment by Olawale1** [comment]
Hi @rohitwaghchaure 

I can confirm it's working perfectly now. Thanks a lot for the great job 👍 

I really hope the refactored code will handle a much more significant number of Customers. I think it might be a good idea to generate a dummy list of customers (say 100,000) for testing purposes so that we don't start experiencing these issues again in a very short while

Thanks
